### PR TITLE
Add flink engine to master

### PR DESCRIPTION
**Pull request by zhangayqian**

### CODE REVIEW

File: engine-flink/src/main/java/org/apache/kylin/engine/flink/FlinkExecutable.java [hunk 1]
```diff
@@ -205,7 +205,7 @@ protected ExecuteResult doWork(ExecutableContext context) throws ExecuteExceptio
                     //flink on yarn specific option (pattern : -yn 1)
                     if (configOptionKey.startsWith("-y") && !entry.getValue().isEmpty()) {
                         sb.append(" ").append(configOptionKey).append(" ").append(entry.getValue());
-                    } else {
+                    } else if(!configOptionKey.startsWith("-y")){
```
Comment (yanghua) [inline engine-flink/src/main/java/org/apache/kylin/engine/flink/FlinkExecutable.java]: Good catch! I remembered there is a bug about the config option. Thanks for your fix!

**Comment by zhangayqian** [comment]
Thank you for your review! I have verify it via CI, and it works well.

**Comment by zhangayqian** [comment]
### Check list
- [ ] Step 1. [**OPTIONAL**]Understand background/root cause/design basically(one hour EST) . Its issue type is 
    - [ ] bug fix
    - [x] new feature
    - [ ] enhancement
- [ ] Step 2. Test cases are designed and documented (30 minutes EST).
- [ ] Step 3. Prepare specific env, for example:
    - mock data (maybe some specific data type)
    - test env (maybe install a RDBMS instance).
- [x] Step 4. Verify and make sure test cases passed.
- [x] Step 5. Paste manual important steps and screenshots here (20 minutes EST). 
    - If you find difficulty in pick most important evidence, please attach diagnosis package. 
- [x] Step 6. Do more check in user perspective (20 minutes EST)
    - [x] Doc need be updated? And if it is updated? Ask help for release manager if so.
    - [x] If it is a breaking change so we should notify Kylin community? Ask help for release manager if so.
- [x] Step 7.  Summarize this test (20 minutes EST).

### Total estimate
- 4 hours for small issue from optimistic estimate. 

### Note
If you find some unexpected and **unrelated** error/mistake, please DO report it if it is truly a mistake, I think we may research and fix it in the future.

If you find background information/root cause analysis is not complete or ambiguous, please try to contact to author or do a quick research and record what you find. It is a good chance to learn something interesting.

If you find it is not easy to design testcase, please notify release manager.

**Comment by zhangayqian** [comment]
## Design test cases

### Prepare test env
- bin/download-flink.sh
- bin/sample.sh
- bin/kylin.sh start

### Case-1
Build segment:
- Modify build engine of `kylin_sales_cube` to FLINK;
- Build two segment `2012-01-01~2012-02-01` and `2012-02-01~2012-03-01`
### Case-2
Merge segment:
- Build two segment `2012-01-01~2012-02-01` and `2012-02-01~2012-03-01`
### Case-3
Query:
- Query-1
```
select PART_DT, sum(price), count(distinct SELLER_ID)
from kylin_sales
join kylin_cal_dt
on kylin_sales.PART_DT = kylin_cal_dt.CAL_DT
where PART_DT > '2012-01-15'
group by PART_DT
```
- Query-2
```
select SELLER_ID, sum(price)
from kylin_sales
group by SELLER_ID
order by sum(price) desc
limit 10
```

**Comment by zhangayqian** [comment]
### Prepare test env
[x] Built on commit_SHA : 5d592b91769b5129c76df3c8b1deb3bd58ebcbb8
[x]Hadoop env : CDH 5.7

**Comment by zhangayqian** [comment]
## Verify test case

### Case-1 (Build failed)
- Build failed in `Convert to hfile` step.
- Error log
image
- Workaround: https://issues.apache.org/jira/browse/KYLIN-4552

**Comment by zhangayqian** [comment]
## Verify test case after fixed
### Case-1 (success)
image

### Case-2 (success)
image


### Case-3 (success)
image

image

**Comment by zhangayqian** [comment]
## Summary
- Happy path test success with FLINK engine.

**Pull request merged**